### ISSUE TITLE
Add per-interaction weights to simulate time decay effects

### ISSUE THREAD

**Issue by FlorianWilhelm** [lyst/lightfm#52]
Quite often more recent interactions of users with items carry more information than long past interactions. For instance the fashion taste of a hipster may change more towards rockabilly over time which will be reflected by his/her interactions. Having the possibility to weight interactions differently, more recent ones higher than older ones, would allow us to model these effects.

**Comment by seanlaw** [comment]
+1

**Comment by maciejkula** [comment]
On it.

**Comment by VaibhavBehl** [comment]
+1 would be a very good feature addition

**Comment by maciejkula** [comment]
I changed the API slightly in https://github.com/lyst/lightfm/pull/61, please have a play and tell me what you think. Accepting a sparse matrix makes it much harder to attach the wrong weights to your interactions.

**Comment by seanlaw** [comment]
Tried this with the implicit feedback section of the MovieLens example:

`model = LightFM(learning_rate=0.05, loss='bpr')
weight = train.copy()
weight.data = np.random.rand(weight.getnnz())
model.fit(train, epochs=10, sample_weight=weight)
train_precision = precision_at_k(model,
                                 train,
                                 10)
test_precision = precision_at_k(model,
                                test,
                                10)

train_auc = full_auc(model, train)
test_auc = full_auc(model, test)
    
print('Precision: %s, %s' % (train_precision, test_precision))
print('AUC: %s, %s' % (train_auc, test_auc))`

I might be doing something wrong but I'm trying to pass random weights and I expected the results to be different when calculating the AUC. The results were the same as the original MovieLens example without weights passed.

**Comment by maciejkula** [comment]
This adds an additional check that everything works as expected: https://github.com/lyst/lightfm/pull/62

Clearly random weights don't make that much difference!

**Comment by seanlaw** [comment]
Awesome! I wonder if it makes a difference if we did an inverse decay where older ratings are weighted higher than recent ones. Just curious.

**Comment by maciejkula** [comment]
I guess you'll know if it's useful when you have a task that requires it. Thanks for your help.

Resolved by #61.
